### ISSUE TITLE
3.0 suggestion: Switch duk_call_method() this/func order

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1345]
Stack signature for `duk_call_method()` is currently:
```
[ ... func this arg1 ... argN ] -> [ ... retval ]
```

This is awkward, as it prevents the following quite natural code from working:
```c
duk_push_global_stash(ctx);
duk_get_prop_string(ctx, -1, "babelCore");
duk_get_prop_string(ctx, -1, "transform");
duk_push_lstring_t(ctx, code_string);
duk_push_object(ctx);
duk_call_method(ctx, 2);

// ...do whatever with the result

duk_pop_2(ctx);
```

In order to make this call, `babelCore` and `transform` need to be swapped after the fact (one is a property of the other, so you can't just reverse the duk_get_prop_string() calls).

**Comment by svaarala** [comment]
This change would be quite awkward (error prone) to upgrade in calling code, at least if the method name stayed the same.

Apart from this it probably wouldn't be a big internal change; instead of the "this" binding being at bottom_index-1, it would be at bottom_index-2.

**Comment by fatcerberus** [comment]
I realized that it's possible to swap the two top values using the idiom `duk_insert(ctx, -2);` but that isn't exactly obvious to someone just quickly skimming the code.

I do see where you're coming from re: upgrading though.

**Comment by svaarala** [comment]
If this change were made (not sure about what I think yet :-) it might be possible to do it at the same time that multiple return values are added. That will require changes to method signatures anyway to indicate how you want the return value(s) -- one return value only, N return values, or return values packed into an array.

So changing the calling convention with that wouldn't break existing call sites, and if the multi-return capable API became the default one, the existing calls could then be left as is for existing code bases (perhaps doing the order swap internally).

**Comment by svaarala** [comment]
Another relevant API call is: http://duktape.org/api.html#duk_swap.

**Comment by fatcerberus** [comment]
Ah, thanks.  I missed that one somehow.

The main problem with `duk_insert()` as an idiom is that, when you encounter the line you have to make the mental leap to understand that the call *also* pops the top value (the one to be inserted).  Just seeing the word "insert" you intuitively think "okay, this adds a new value" and so it's maybe a bit confusing.  `duk_swap()` indeed makes it clearer what's going on.

**Comment by svaarala** [comment]
Agree, and with `duk_insert()` it's also important to understand all indices are before popping the argument (which is generally the API policy, i.e. indices are before any changes).
